include/irr/asset/material_compiler/CMaterialCompilerGLSLBackendCommon.h

```diff
@@ -0,0 +1,506 @@
+			core::bitfieldExtract(i, BITFIELDS_BSDF_BUF_OFFSET_SHIFT, BITFIELDS_BSDF_BUF_OFFSET_WIDTH)
+		);
+	}
+
+	inline void setBSDFDataIx(instr_t& i, uint32_t ix)
+	{
+		i = core::bitfieldInsert<instr_t>(i, ix, BITFIELDS_BSDF_BUF_OFFSET_SHIFT, BITFIELDS_BSDF_BUF_OFFSET_WIDTH);
+	}
+
+	// more like max param number plus one (includes dummies)
+	inline uint32_t getParamCount(E_OPCODE op)
+	{
+		switch (op)
+		{
+		case OP_DIFFUSE: [[fallthrough]];
+		case OP_PLASTIC: [[fallthrough]];
+		case OP_COATING:
+			return 4u;
+		case OP_CONDUCTOR: [[fallthrough]];
+		case OP_DIFFTRANS: [[fallthrough]];
+		case OP_DIELECTRIC: [[fallthrough]];
+		case OP_THINDIELECTRIC:
+			return 3u;
+		case OP_BLEND: [[fallthrough]];
+		case OP_BUMPMAP:
+			return 1u;
+		default: 0u;
+		}
+	}
+
+	using VTID = asset::ICPUVirtualTexture::SMasterTextureData;
+#include "irr/irrpack.h"
+	struct STextureData {
+		VTID vtid = VTID::invalid();
+		union {
+			uint32_t prefetch_reg;//uint32
+			uint32_t scale = 0u;//float
+		};
+
+		bool operator==(const STextureData& rhs) const { return memcmp(this,&rhs,sizeof(rhs))==0; }
+		struct hash
+		{
+			std::size_t operator()(const STextureData& t) const { return std::hash<uint64_t>{}(reinterpret_cast<const uint64_t&>(t.vtid)) ^ std::hash<uint32_t>{}(t.scale); }
+		};
+	} PACK_STRUCT;
+
+	struct STextureOrConstant
+	{
+		void setConst(float f) { std::fill(constant, constant+3, reinterpret_cast<uint32_t&>(f)); }
+		void setConst(const float* fv) { memcpy(constant, fv, sizeof(constant)); }
+		void setTexture(const VTID& _vtid, float _scale)
+		{
+			tex.vtid = _vtid;
+			core::uintBitsToFloat(tex.scale) = _scale;
+		}
+		core::vector3df_SIMD getConst() const
+		{
+			auto ret = core::vector3df_SIMD(reinterpret_cast<const float*>(constant));
+			ret.w = 0.f;
+			return ret;
+		}
+
+		union
+		{
+			STextureData tex;
+			uint32_t constant[3];//3x float
+		};
+	} PACK_STRUCT;
+#include "irr/irrunpack.h"
+
+#include "irr/irrpack.h"
+	struct SAllDiffuse
+	{
+		STextureOrConstant alpha;
+		STextureOrConstant dummy;
+		STextureOrConstant opacity;
+		STextureOrConstant reflectance;
+	} PACK_STRUCT;
+	struct SDiffuseTransmitter
+	{
+		STextureOrConstant transmittance;
+		STextureOrConstant dummy;
+		STextureOrConstant opacity;
+	} PACK_STRUCT;
+	struct SAllDielectric
+	{
+		STextureOrConstant alpha_u;
+		STextureOrConstant alpha_v;
+		STextureOrConstant opacity;
+		STextureOrConstant dummy;
+		float eta;
+	} PACK_STRUCT;
+	struct SAllConductor
+	{
+		STextureOrConstant alpha_u;
+		STextureOrConstant alpha_v;
+		STextureOrConstant opacity;
+		//3d complex IoR, rgb19e7 format, [0]=real, [1]=imaginary
+		uint64_t eta[2];
+	} PACK_STRUCT;
+	struct SAllPlastic
+	{
+		STextureOrConstant alpha_u;
+		STextureOrConstant alpha_v;
+		STextureOrConstant opacity;
+		STextureOrConstant reflectance;
+		float eta;
+	} PACK_STRUCT;
+	struct SAllCoating
+	{
+		STextureOrConstant alpha_u;
+		STextureOrConstant alpha_v;
+		STextureOrConstant opacity;
```

Comment:
does `opacity` override thickness+sigma beer's law thing?

Reply:
No idea, i dont know a thing how to implement coating

Reply:
what does mitsuba say?